### DISCUSSION TITLE
"Signed request date outside acceptable time window" when setting HTTP 'Date' header

### DISCUSSION THREAD

**Discussion by sventiffe** [mastodon/mastodon#21588] [Q&A] [answered]
I'm trying to create signed PUT requests in Java using a `HttpURLConnection conn = (HttpURLConnection) url.openConnection();`. One problem I seem to have is that I cannot set the date manually in the request, although being needed for the signed headers.

If I manually set the date header of an signed HTTP request like
```
 conn.setRequestProperty("Date", "24 Nov 2022 21:40:36 UTC");
```
the server will return ` Signed request date outside acceptable time window`. This error message seems to be created [here](https://github.com/mastodon/mastodon/blob/75bbe3b1b80207d9cab1a73a9fc38324ee6340d5/app/controllers/concerns/signature_verification.rb#L84) in ` signed_request_actor` and if I read [matches_time_window](https://github.com/mastodon/mastodon/blob/75bbe3b1b80207d9cab1a73a9fc38324ee6340d5/app/controllers/concerns/signature_verification.rb#L200) the reason why the error is not thrown when I do not add it, it's because it would return true if created_time remains nil.

If instead I set the header Date, it will assign
```
 created_time = Time.httpdate(request.headers['Date']).utc
```
As I am not using `expires` as signing header, I assume `expires_time` stays nil and is then computed as the header Date plus five minutes:
```
expires_time ||= created_time + 5.minutes unless created_time.nil?
```
Now that both `created_time` and `expires_time` are present, these criteria should not apply:
```
return false if created_time.present? && created_time > Time.now.utc + CLOCK_SKEW_MARGIN
return false if expires_time.present? && Time.now.utc > expires_time + CLOCK_SKEW_MARGIN
```
`created_time` should be less than Time.now.utc (plus one hour). `now` should be smaller than `expires_time` (plus that one hour which I do not fully understand).

So it should return true.  Where is my mistake? Is my date format ` "24 Nov 2022 21:40:36 UTC"` an issue?

**Comment by nightpool** [comment, marked as answer]
Yes, see the specs for the HTTP Date header here: https://httpwg.org/specs/rfc9110.html#field.date and the syntax definition for this header here: https://httpwg.org/specs/rfc9110.html#http.date

Here are some examples of correct formatting for the HTTP Date header:

> `Sun, 06 Nov 1994 08:49:37 GMT    ; IMF-fixdate`
> 
> Examples of the two obsolete formats are
> 
> `Sunday, 06-Nov-94 08:49:37 GMT   ; obsolete RFC 850 format`
> `Sun Nov  6 08:49:37 1994         ; ANSI C's asctime() format`

I recommend you find a library to handle HTTP protocol concerns for you rather then trying to implement them yourself.

**Reply by sventiffe** [reply]
Thanks. For completeness, DateTimeFormatter.RFC_1123_DATE_TIME works.

**Reply by sventiffe** [reply]
For posterity, DateTimeFormatter.RFC_1123_DATE_TIME [will fail](https://bugs.openjdk.org/browse/JDK-8075908) for days-in-month below 10 as it won't use a leading 0 for the day as expected. This formatter works for IMF-fixdate format: https://jackrabbit.apache.org/api/2.20/org/apache/jackrabbit/webdav/util/HttpDateTimeFormatter.html
